### PR TITLE
[13.0][ADD] stock_picking_mixin, stock_picking_sale_state, stock_picking_purchase_state

### PR DESCRIPTION
**Pull request by ChristianSantamaria**
Add the Picking Status field to Sale.Order, Purchase.Order and its tree.

### CODE REVIEW

File: stock_picking_purchase_state/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from . import models
```
Comment (dalonsod) [inline stock_picking_purchase_state/__init__.py]: Missing EOL

File: stock_picking_purchase_state/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        Add the Picking Status field to Purchase.Order and its tree.
+    """,
+    "author": "Solvos",
+    "license": "LGPL-3",
+    "version": "13.0.1.0.0",
+    "category": "stock, purchase",
+    "website": "https://github.com/solvosci/slv-stock",
+    "depends": [
+        "stock",
+        "purchase"
```
Comment (dalonsod) [inline stock_picking_purchase_state/__manifest__.py]: Suggested change:
```suggestion
        "purchase",
```

File: stock_picking_purchase_state/models/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+from . import purchase_order
```
Comment (dalonsod) [inline stock_picking_purchase_state/models/__init__.py]: Missing EOL

File: stock_picking_purchase_state/models/purchase_order.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+            " * Done: The transfer has been processed.\n"
+            " * Cancelled: The transfer has been cancelled.")
+
+    @api.depends('picking_ids.state')
+    def _compute_picking_status(self):
+        for order in self:
+            if order.picking_ids.filtered(lambda p: p.state != "cancel"):
+                order.picking_status = order.picking_ids.filtered(lambda p: p.state != "cancel").sorted(key="scheduled_date")[-1].state
+            else:
+                order.picking_status = False
```
Comment (dalonsod) [inline stock_picking_purchase_state/models/purchase_order.py]: Proposed refactoring in order to reduce line character count:
```suggestion
            pickings = order.picking_ids.filtered(
                lambda p: p.state != "cancel"
            )
            if pickings:
                order.picking_status = pickings.sorted(
                    key="scheduled_date"
                )[-1].state
            else:
                order.picking_status = False
```

File: stock_picking_sale_state/__init__.py [hunk 5]
```diff
@@ -0,0 +1 @@
+from . import models
```
Comment (dalonsod) [inline stock_picking_sale_state/__init__.py]: Missing EOF

**Review by dalonsod — CHANGES_REQUESTED** [review]
Only code review, and for the `purchase` addon, see comments and also apply, if possible, to the `sale` one. More requirements:
- could you add to purchase order `search` view the new field to the `field` and `groupby` list?
- @lmiguens-solvos should we need this new field also in `form` view of an order?

**Comment by lmiguens-solvos** [comment]
> * @lmiguens-solvos should we need this new field also in `form` view of an order?

No, I think it is enough to see it in the tree view (but remember, both tree view, view_order_tree and view_quotation_tree)

File: stock_picking_mixin/__manifest__.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# © 2021 Solvos Consultoría Informática (<http://www.solvos.es>)
+# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
+{
+    "name": "Stock Picking Mixin",
```
Comment (dalonsod) [inline stock_picking_mixin/__manifest__.py]: Preventing usage for other purposes, I think is better renaming it
```suggestion
    "name": "Stock Picking Mixins",
```

and, then rename the whole addon as `stock_picking_mixins`.

File: stock_picking_mixin/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    "summary": """
+        Add the Picking Status field.
```
Comment (dalonsod) [inline stock_picking_mixin/__manifest__.py]: Update summary with proper description

File: stock_picking_mixin/models/stock_picking_mixin.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
+# © 2021 Solvos Consultoría Informática (<http://www.solvos.es>)
+# License LGPL-3.0 (http://www.gnu.org/licenses/lgpl-3.0.html)
+
+from odoo import fields, api, models
+
+
+class StockPickingMixin(models.AbstractModel):
+    _name = "stock.picking.mixin"
```
Comment (dalonsod) [inline stock_picking_mixin/models/stock_picking_mixin.py]: The current abstract model should be renamed, and the functionality isolated within this addon:
```suggestion
    _name = "stock.picking.state.mixin"
```
Then, adding new abstract functionalities to this addon becomes adding new and independent models

File: stock_picking_mixin/readme/DESCRIPTION.rst [hunk 9]
```diff
@@ -0,0 +1 @@
+Add the Picking Status field.
```
Comment (dalonsod) [inline stock_picking_mixin/readme/DESCRIPTION.rst]: Change description

File: stock_picking_purchase_state/__manifest__.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
+{
+    "name": "Stock Picking Purchase State",
+    "summary": """
+        Add the Picking Status field to Purchase.
+    """,
+    "author": "Solvos",
+    "license": "LGPL-3",
+    "version": "13.0.1.0.0",
+    "category": "stock, purchase",
```
Comment (dalonsod) [inline stock_picking_purchase_state/__manifest__.py:11]: Does it work? I mean using multiple categories for an addon description

**Review by dalonsod — CHANGES_REQUESTED** [review]
Only code review, see comments

File: stock_picking_purchase_state/__manifest__.py (hunk 10 above)
Comment (ChristianSantamaria) [reply stock_picking_purchase_state/__manifest__.py:11]: Yes it works, for example in a [stock-logistics-workflow ](https://github.com/OCA/stock-logistics-workflow/blob/acf795670aa5bfdcee32ba1ee526194f866a2951/stock_no_negative/__manifest__.py#L9)addon it has three categories

File: stock_picking_mixins/__manifest__.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
+# © 2021 Solvos Consultoría Informática (<http://www.solvos.es>)
+# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
+{
+    "name": "Stock Picking Mixins",
+    "summary": """
+        Adds a structure to insert the Picking Status field for purchase and sale orders.
```
Comment (dalonsod) [inline stock_picking_mixins/__manifest__.py]: I'd prefer a more generic summary, like this:
```suggestion
        Adds generic structures related to stock picking models
```

File: stock_picking_mixins/readme/DESCRIPTION.rst [hunk 12]
```diff
@@ -0,0 +1 @@
+Add the Picking Status field.
```
Comment (dalonsod) [inline stock_picking_mixins/readme/DESCRIPTION.rst]: The same as above

File: stock_picking_mixins/models/stock_picking_mixins.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
+# © 2021 Solvos Consultoría Informática (<http://www.solvos.es>)
+# License LGPL-3.0 (http://www.gnu.org/licenses/lgpl-3.0.html)
+
+from odoo import fields, api, models
+
+
+class StockPickingMixins(models.AbstractModel):
+    _name = "stock.picking.state.mixins"
```
Comment (dalonsod) [inline stock_picking_mixins/models/stock_picking_mixins.py]: This is lonely mixin, so should be like this:
```suggestion
class StockPickingMixin(models.AbstractModel):
    _name = "stock.picking.state.mixin"
```

- The addon contains one mixin for now, but it should allocate more models (mixins) in the future => mixins
- Every model should represent  a mixin  => mixin

File: stock_picking_purchase_state/__manifest__.py [hunk 14]
```diff
@@ -0,0 +1,21 @@
+# © 2021 Solvos Consultoría Informática (<http://www.solvos.es>)
+# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
+{
+    "name": "Stock Picking Purchase State",
+    "summary": """
+        Add the Picking Status field to Purchase.
```
Comment (dalonsod) [inline stock_picking_purchase_state/__manifest__.py]: Suggested change:
```suggestion
        Adds Picking Status field to Purchase Orders.
```

File: stock_picking_sale_state/__manifest__.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
+# © 2021 Solvos Consultoría Informática (<http://www.solvos.es>)
+# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
+{
+    "name": "Stock Picking Sale State",
+    "summary": """
+        Add the Picking Status field to Sale.
```
Comment (dalonsod) [inline stock_picking_sale_state/__manifest__.py]: Suggested change:
```suggestion
        Adds Picking Status field to Sale Orders.
```

**Review by dalonsod — CHANGES_REQUESTED** [review]
Still code review, only minor changes. When attended, I'll test it

**Comment by dalonsod** [comment]
Locally tested,
- Sales: once confirmed an order, state is properly updated. Then, cancelled the current picking, the state is unset. Finally, create another picking, the state is updated again => Everything is OK
- Purchase: made the same test, it didn't work, but the problem seems to be in Odoo itself, because manually creating a new picking document within a purchase order does not actually link it to the current order (despite "source document" is properly filled). @lmiguens-solvos have you detected this issue in Odoo?

**Comment by dalonsod** [comment]
@ChristianSantamaria as you warned, `es` translation is not applying, tested on my environment.

I really don't know if the problem is that, but try to remove `picking_status` `help` attribute, and then remove also the related translations. It's a complex translation, maybe it has an issue and the other terms are affected within the file

File: stock_picking_purchase_state/models/purchase_order.py [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+# License LGPL-3.0 (http://www.gnu.org/licenses/lgpl-3.0.html)
+
+from odoo import models, fields
+
+
+class PurchaseOrder(models.Model):
+    _name = "purchase.order"
+    _inherit = ["purchase.order", "stock.picking.state.mixin"]
+    
+    picking_status = fields.Selection([
```
Comment (dalonsod) [inline stock_picking_purchase_state/models/purchase_order.py:13]: Plese add here a comment referring to the Odoo known bug with mixins and translations, including the related URL

File: stock_picking_sale_state/models/sale_order.py [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+# License LGPL-3.0 (http://www.gnu.org/licenses/lgpl-3.0.html)
+
+from odoo import models, fields
+
+
+class SaleOrder(models.Model):
+    _name = "sale.order"
+    _inherit = ["sale.order", "stock.picking.state.mixin"]
+    
+    picking_status = fields.Selection([
```
Comment (dalonsod) [inline stock_picking_sale_state/models/sale_order.py:13]: The same as above

**Review by dalonsod — CHANGES_REQUESTED** [review]
Tested, LGTM. Please attend comments and squash commits

**Comment by lmiguens-solvos** [comment]
> * Purchase: made the same test, it didn't work, but the problem seems to be in Odoo itself, because manually creating a new picking document within a purchase order does not actually link it to the current order (despite "source document" is properly filled). @lmiguens-solvos have you detected this issue in Odoo?

I can't find anything ... It's true that I don't really know how to search it either ...

**Review by dalonsod — APPROVED** [review]

**Comment by dalonsod** [comment]
@lmiguens-solvos related problem with picking addition to a purchase with existing pickings is probably not linked with this addon, so I've approved it. @lmiguens-solvos when your approval arrives I'll merge it

**Comment by lmiguens-solvos** [comment]
> @lmiguens-solvos when your approval arrives I'll merge it

Just 2 details: 
- the translation of Ready in the picking state is "Preparado" instead of "Listo". It is better that it translates the same.
- and this new field picking_status is only in quotation_tree. I think it should also add in the view_order_tree. 
what do you think, @dalonsod ?? 
There are two views diferents (both in sale and purchase) in "Orders" menu: "Quotations" (the view you modified) and "Orders" (the other view that I propose to modify as well)

**Comment by dalonsod** [comment]
> and this new field picking_status is only in quotation_tree. I think it should also add in the view_order_tree.
what do you think, @dalonsod ??
There are two views diferents (both in sale and purchase) in "Orders" menu: "Quotations" (the view you modified) and "Orders" (the other view that I propose to modify as well)

you're right, I've only tested from purchase quotations. In this view you can actually see all ordered purchases, it's a bit strange. Picking status has no sense for quotations, but I'd preserve it. In Sales structure is a little bit different, but appear to be the same effect: from "quotations" view we should consider picking status, because orders are also shown

**Review by dalonsod — CHANGES_REQUESTED** [review]
Locally tested, grouping by Picking Status in purchase orders (not quotations) seems to be missing. Sale orders & quotations OK

**Review by dalonsod — APPROVED** [review]
LGTM. Please squash commits again and I'll merge it

**Review by lmiguens-solvos — APPROVED** [review]
LGTM.
For a moment I wondered why you consider the last pick ordering by scheduled_date. But, I guess you decided for something.

**Pull request merged**